contracts/Governance/Timelock.sol

```diff
@@ -1,123 +1,174 @@
+    }
 
-    address public admin;
-    address public pendingAdmin;
-    uint256 public delay;
+    uint256 public minimumExecutionDelay;
+    uint256 public maximumExecutionDelay;
+    uint256 public executionGracePeriod;
+
+    address public proposer;
+    address public executor;
+
+    mapping(bytes32 => TransactionStatus) public queuedTransactions;
+
+    event QueuedTransactionsBatch(bytes32 indexed txParamsHash, uint256 numberOfTransactions, uint256 eta);
+    event CancelledTransactionsBatch(bytes32 indexed txParamsHash);
+    event ApprovedTransactionsBatch(bytes32 indexed txParamsHash);
+    event ExecutedTransaction(bytes32 indexed txParamsHash, address indexed recipient, uint256 value, bytes data, uint256 eta);
+    event NewProposerSet(address indexed previousProposer, address indexed newProposer);
+    event NewExecutorSet(address indexed previousExecutor, address indexed newExecutor);
+    event NewExecutionDelaySet(uint256 indexed delay);
+    event NewMinimumExecutionDelaySet(uint256 indexed delay);
+    event NewMaximumExecutionDelaySet(uint256 indexed delay);
+    event NewExecutionGracePeriodSet(uint256 indexed gracePeriod);
+
+    constructor(
+        address _proposer,
+        address _executor,
+        uint256 _minimumExecutionDelay,
+        uint256 _maximumExecutionDelay,
+        uint256 _executionGracePeriod
+    ) {
+        require(_minimumExecutionDelay <= _maximumExecutionDelay, "Timelock: The minimum delay cannot be larger than the maximum execution delay");
+
+        setMinimumExecutionDelay(_minimumExecutionDelay);
+        setMaximumExecutionDelay(_maximumExecutionDelay);
+        setExecutionGracePeriod(_executionGracePeriod);
+        setProposer(_proposer);
+        setExecutor(_executor);
+    }
 
-    mapping(bytes32 => bool) public queuedTransactions;
+    function setProposer(address _proposer) public onlyOwner {
+        require(_proposer != address(0), "Timelock: Proposer address cannot be 0");
 
-    constructor(address admin_, uint256 delay_) {
-        require(admin_ != address(0), "Admin address cannot be 0");
-        require(delay_ >= MINIMUM_DELAY, "Timelock::constructor: Delay must exceed minimum delay.");
-        require(delay_ <= MAXIMUM_DELAY, "Timelock::setDelay: Delay must not exceed maximum delay.");
+        address previousProposer = proposer;
+        proposer = _proposer;
 
-        admin = admin_;
-        delay = delay_;
+        emit NewProposerSet(previousProposer, proposer);
     }
 
-    function setDelay(uint256 delay_) external {
-        require(msg.sender == address(this), "Timelock::setDelay: Call must come from Timelock.");
-        require(delay_ >= MINIMUM_DELAY, "Timelock::setDelay: Delay must exceed minimum delay.");
-        require(delay_ <= MAXIMUM_DELAY, "Timelock::setDelay: Delay must not exceed maximum delay.");
-        delay = delay_;
+    function setExecutor(address _executor) public onlyOwner {
+        address previousExecutor = executor;
```

Comment:
Let me know what's the decision :)

Reply:
@kazazor What is the risk here?
Since it's an easy change (correct me if I'm wrong) I think we should change it to a map, just in case we will need it in the future.

Reply:
The problem with a map is that you can't really see what's the current values there.. So if someone will add him/herself there, will be hard for us to catch that.
Maybe using an enumerable set would be a better fit here? because then we can also list it...

Reply:
done